ffi/examples/read-table/read_table.c

```diff
@@ -0,0 +1,145 @@
+#include "arrow.h"
+#endif
+
+struct EngineContext {
+  GlobalScanState *global_state;
+  const ExternEngineInterfaceHandle *engine_interface;
+};
+
+void print_selection_vector(char* indent, const struct KernelBoolSlice *selection_vec) {
+  for (int i = 0; i < selection_vec->len; i++) {
+    printf("%ssel[%i] = %b\n", indent, i, selection_vec->ptr[i]);
+  }
+}
+
+void* allocate_string(const struct KernelStringSlice slice) {
+  char* buf = malloc(sizeof(char) * (slice.len + 1)); // +1 for null
+  snprintf(buf, slice.len + 1, "%s", slice.ptr);
+  return buf;
+}
+
+void visit_callback(void* engine_context, const struct KernelStringSlice path, long size, struct CDvInfo *dv_info, struct CStringMap *partition_values) {
+  printf("called back to actually read!\n  path: %.*s\n", path.len, path.ptr);
+  struct EngineContext *context = engine_context;
+  KernelBoolSlice *selection_vector = selection_vector_from_dv(dv_info, context->engine_interface, context->global_state);
+  if (selection_vector) {
+    printf("  Selection vector:\n");
+    print_selection_vector("    ", selection_vector);
+    drop_bool_slice(selection_vector);
+  } else {
+    printf("  No selection vector for this call\n");
+  }
+  // normally this would be picked out of the schema
+  char* letter_key = "letter";
+  KernelStringSlice key = {letter_key, strlen(letter_key)};
+  char* partition_val = get_from_map(partition_values, key, allocate_string);
+  if (partition_val) {
+    printf("  letter partition here: %s\n", partition_val);
+    free(partition_val);
+  } else {
+    printf("  no partition here\n");
+  }
+}
+
+void visit_data(void *engine_context, struct EngineDataHandle *engine_data, const struct KernelBoolSlice selection_vec) {
+  printf("Got some data\n");
+  printf("  Of this data, here is a selection vector\n");
+  print_selection_vector("    ", &selection_vec);
+  visit_scan_data(engine_data, selection_vec, engine_context, visit_callback);
+}
+
+int main(int argc, char* argv[]) {
+  if (argc < 2) {
+    printf("Usage: %s table/path\n", argv[0]);
+    return -1;
+  }
+
+  char* table_path = argv[1];
+  printf("Reading table at %s\n", table_path);
+
+  KernelStringSlice table_path_slice = {table_path, strlen(table_path)};
+
+  ExternResultEngineInterfaceBuilder interface_builder_res =
+    get_engine_interface_builder(table_path_slice, NULL);
+  if (interface_builder_res.tag != OkEngineInterfaceBuilder) {
+    printf("Could not get engine interface builder\n");
+    return -1;
+  }
+
+  // an example of using a builder to set options when building a engine interface
+  struct EngineInterfaceBuilder *interface_builder = interface_builder_res.ok;
+  char* key_test = "key1";
+  KernelStringSlice key_slice = {key_test, strlen(key_test)};
+  char* val_test = "val1";
+  KernelStringSlice val_slice = {val_test, strlen(val_test)};
+  set_builder_option(interface_builder, key_slice, val_slice);
+  ExternResultExternEngineInterfaceHandle engine_interface_res =
```

Comment:
Rescuing comment about naming:
>  Some judicious typedef could really help readability... or maybe we just switch to C++?

I think we at least _want_ to keep the C stuff usable as well. I've turned off name mangling for now, so the horrible `_____` s are gone. The names are still long, but a little less crazy.

We can iterate on substitutions or a standard set of typedefs we want for them.